tests/queries/0_stateless/02346_text_index_direct_read.sql

```diff
@@ -0,0 +1,108 @@
+
+INSERT INTO tab VALUES (101, 'Alick a01'),
+                       (102, 'Blick a02');
+
+
+----------------------------------------------------
+SELECT 'Test hasToken';
+
+SELECT count() FROM tab WHERE hasToken(text, 'Alick');
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time DESC LIMIT 1;
+
+----------------------------------------------------
+SELECT 'Test searchAll';
+
+SELECT count() FROM tab WHERE searchAll(text, ['Alick']);
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+----------------------------------------------------
+SELECT 'Test searchAny';
+
+SELECT count() FROM tab WHERE searchAny(text, ['Alick']);
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+----------------------------------------------------
+-- Adds column, but keeps the original
+SELECT 'Test hasToken + length(text)';
+
+SELECT count() FROM tab WHERE hasToken(text, 'Alick') or length(text) > 1;
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+----------------------------------------------------
+-- Adds column, but keeps the original
+SELECT 'Test select text + hasToken';
+
+SELECT text FROM tab WHERE searchAny(text, ['Alick']);
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+
+----------------------------------------------------
+-- 2 substitutions
+----------------------------------------------------
+SELECT 'Test hasToken and hasToken';
+
+SELECT count() FROM tab WHERE hasToken(text, 'Alick') and hasToken(text, 'Blick');
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+----------------------------------------------------
+SELECT 'Test searchAny or hasToken';
+
+SELECT count() FROM tab WHERE searchAny(text, ['Blick']) or hasToken(text, 'Alick');
+
+SYSTEM FLUSH LOGS text_log;
+
+SELECT message FROM system.text_log WHERE
+       (logger_name = 'optimizeDirectReadFromTextIndex') and
+       startsWith(message, 'Added:') ORDER BY event_time_microseconds DESC LIMIT 1;
+
+
+DROP TABLE IF EXISTS tab;
```

Comment:
While at it, could you please extend the "direct read" section in the [docs](https://clickhouse.com/docs/engines/table-engines/mergetree-family/invertedindexes#direct-read)? It needs
- a high-level description of direct read,
- a list of supported functions,
- a note how it plays together with AND/OR/NOT and/or other (potentially non-text search) functions.

Reply:
Solved via:  dbc38e42988e29cae9ffb79d1e

Please specify if you prefer to add more examples or more details in the text.